### PR TITLE
[MIG]port cagnotte modules to V10.0

### PR DESCRIPTION
**Pull request by Cedric-Pigeon**

### CODE REVIEW

File: cagnotte_base/models/cagnotte.py [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+                cagnotte_type = self.env['cagnotte.type'].search(
+                    [('account_id', '=', values['account_id'])])
+                if cagnotte_type:
+                    # create cagnotte
+                    values['account_cagnotte_id'] = \
+                        self.env['account.cagnotte'].create(
+                            {'cagnotte_type_id': cagnotte_type.id}).id
+        return values
+
+    @api.multi
```
Comment (rousseldenis) [inline cagnotte_base/models/cagnotte.py]: @Cedric-Pigeon Careful, this must be api.model

**Review by rousseldenis — CHANGES_REQUESTED** [review]
Code review, Seems pretty good to me

File: cagnotte_base/__manifest__.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    'author': "ACSONE SA/NV",
+    "category": "Accounting & Finance",
+    "website": "http://www.acsone.eu",
+    "depends": [
+        "account",
+    ],
+    "data": [
+        "views/cagnotte_views.xml",
+        "security/ir.model.access.csv",
+        "security/cagnotte_base_security.xml",
```
Comment (JonathanNEMRY) [inline cagnotte_base/__manifest__.py:16]: more preferable to load security before case of using some groups into views

File: cagnotte_base/demo/account_cagnotte_demo.xml [hunk 3]
```diff
@@ -0,0 +1,43 @@
+<?xml version="1.0" encoding="utf-8"?>
+<openerp>
```
Comment (JonathanNEMRY) [inline cagnotte_base/demo/account_cagnotte_demo.xml]: you can rename this one into odoo

File: cagnotte_base/demo/account_cagnotte_demo.xml [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    <data>
```
Comment (JonathanNEMRY) [inline cagnotte_base/demo/account_cagnotte_demo.xml]: and remove this one if it is not a noupdate record

File: cagnotte_base/models/cagnotte.py [hunk 5]
```diff
@@ -0,0 +1,202 @@
+# -*- coding: utf-8 -*-
+# © 2015  Laetitia Gangloff, Acsone SA/NV (http://www.acsone.eu)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import api, fields, models
+
+
+class CagnotteType(models.Model):
+    _name = 'cagnotte.type'
+
```
Comment (JonathanNEMRY) [inline cagnotte_base/models/cagnotte.py:14]: please add a _description

File: cagnotte_base/demo/account_cagnotte_demo.xml (hunk 3 above)
Comment (JonathanNEMRY) [inline cagnotte_base/demo/account_cagnotte_demo.xml]: odoo

File: cagnotte_coupon/models/cagnotte.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+        return "%s%s" % (uid, base)
+
+    @api.multi
+    def create(self, vals):
+        res = super(AccountCagnotte, self).create(vals)
+        if not vals.get('coupon_code') and vals.get('cagnotte_type_id'):
+            cagnotte_type = self.env['cagnotte.type'].browse(
+                vals['cagnotte_type_id'])
+            if cagnotte_type.with_coupon_code:
+                res.sudo().coupon_code = self.generate_coupon_code(res.id)
```
Comment (JonathanNEMRY) [inline cagnotte_coupon/models/cagnotte.py:46]: why sudo required here? is the create action triggered by someone that has not  write access?

File: cagnotte_base/models/cagnotte.py [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+    @api.multi
+    @api.depends('account_move_line_ids.debit',
+                 'account_move_line_ids.credit')
+    def _compute_solde_cagnotte(self):
+        for cagnotte in self:
+            solde_cagnotte = 0
+            for move_line in cagnotte.account_move_line_ids:
+                solde_cagnotte += move_line.credit - move_line.debit
+            cagnotte.solde_cagnotte = solde_cagnotte
```
Comment (ThomasBinsfeld) [inline cagnotte_base/models/cagnotte.py]: @Cedric-Pigeon Performances would be better with a read_group.

```python
        @api.multi
        @api.depends('account_move_line_ids.debit',
                                 'account_move_line_ids.credit')
        def _compute_solde_cagnotte(self):
        aml_model = self.env['account.move.line']
        domain = [
            ('account_cagnotte_id', 'in', self.ids)]

        res = aml_model.read_group(
            domain=domain,
            fields=['account_cagnotte_id', 'credit', 'debit'],
            groupby=['account_cagnotte_id']
        )

        aml_dict = {}
        for dic in res:
            cagnotte_id = dic['account_cagnotte_id'][0]
            solde_cagnotte = dic['credit'] - dic['debit']
            aml_dict[cagnotte_id] = solde_cagnotte

        for cagnotte in self:
            cagnotte.solde_cagnotte = aml_dict.get(cagnotte.id, 0.0)
```

But `solde_cagnotte` is stored so no big deal I guess...

File: cagnotte_base/models/cagnotte.py [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+    @api.multi
+    def _check_cagnotte_account(self):
+        """ Account must correspond to cagnotte account
+        """
+        for line in self:
+            if line.account_cagnotte_id:
+                if line.account_cagnotte_id.cagnotte_type_id.account_id.id \
+                        != line.account_id.id:
+                    return False
+        return True
```
Comment (ThomasBinsfeld) [inline cagnotte_base/models/cagnotte.py:189]: @Cedric-Pigeon That could be written:

```python
@api.multi
def _check_cagnotte_account(self):
 """ Account must correspond to cagnotte account
 """
     return not any(l.account_cagnotte_id and l.account_cagnotte_id.cagnotte_type_id.account_id != l.account_id for l in self)
```

File: cagnotte_base/models/cagnotte.py [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+        readonly=True,
+        ondelete='restrict',
+        help="Use this field to give coupon to a customer",
+        states={'draft': [('readonly', False)]}
+    )
+
+    @api.onchange("cagnotte_type_id")
+    def onchange_cagnotte_type_id(self):
+        if self.cagnotte_type_id:
+            self.account_id = self.cagnotte_type_id.account_id.id
```
Comment (ThomasBinsfeld) [inline cagnotte_base/models/cagnotte.py]: @Cedric-Pigeon `.id` is not useful

File: cagnotte_base/tests/data/account_cagnotte_data.xml [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+        </record>
+        <record id="cagnotte_type" model="cagnotte.type">
+            <field name="name">Cagnotte Type</field>
+            <field name="sequence_id" ref="sequence_cagnotte_type"/>
+            <field name="account_id" ref="cagnotte"/>
+            <field name="journal_id" ref="cagnotte_journal"/>
+            <field name="product_id" ref="product_product_cagnotte"/>
+        </record>
+    </data>
+</odoo>
```
Comment (ThomasBinsfeld) [inline cagnotte_base/tests/data/account_cagnotte_data.xml]: @Cedric-Pigeon Newline :scream:

File: cagnotte_partner/views/cagnotte_views.xml [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+            <field name="domain">[('partner_id', '=', active_id)]</field>
+        </record>
+
+        <record id="res_partner_view_form_inherit_cagnotte_partner" model="ir.ui.view">
+            <field name="name">res.partner.form</field>
+            <field name="model">res.partner</field>
+            <field name="inherit_id" ref="base.view_partner_form"/>
+            <field name="priority">60</field>
+            <field name="arch" type="xml">
+                <xpath expr="//div[@name='button_box']" position="inside">
```
Comment (ThomasBinsfeld) [inline cagnotte_partner/views/cagnotte_views.xml]: @Cedric-Pigeon Could be written:

```xml
<div name="button_box" position="inside">
```

File: cagnotte_base/models/cagnotte.py [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+                default_debit_account_id.id
+            if journal_debit_account_id != cagnotte.account_id.id:
+                return False
+            journal_credit_account_id = cagnotte.journal_id.\
+                default_credit_account_id.id
+            if journal_credit_account_id != cagnotte.account_id.id:
+                return False
+        return True
+
+    _constraints = [
```
Comment (ThomasBinsfeld) [inline cagnotte_base/models/cagnotte.py]: @Cedric-Pigeon Should be replaced by `@api.constrains`

File: cagnotte_base/models/cagnotte.py [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+        string='Cagnotte',
+        ondelete='restrict')
+
+    @api.model
+    def cagnotte_value(self, values):
+        """ If used account is on a cagnotte type,
+            create a cagnotte
+        """
+        if not values.get('account_cagnotte_id'):
+            if values.get('account_id') and values.get('credit', 0) > 0:
```
Comment (ThomasBinsfeld) [inline cagnotte_base/models/cagnotte.py]: @Cedric-Pigeon `credit` is a float and should be compared with `float_compare`.

File: cagnotte_base/models/cagnotte.py [hunk 14]
```diff
@@ -0,0 +1,202 @@
...
+        """ Account must correspond to cagnotte account
+        """
+        for line in self:
+            if line.account_cagnotte_id:
+                if line.account_cagnotte_id.cagnotte_type_id.account_id.id \
+                        != line.account_id.id:
+                    return False
+        return True
+
+    _constraints = [
```
Comment (ThomasBinsfeld) [inline cagnotte_base/models/cagnotte.py]: @Cedric-Pigeon Should be replaced by `@api.constrains` and a `validationError` in the method.

File: cagnotte_partner/models/cagnotte.py [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+    @api.multi
+    def _check_partner(self):
+        """ Check there is no move lines to be able to set a partner
+        """
+        for cagnotte in self:
+            if cagnotte.partner_id and cagnotte.account_move_line_ids:
+                return False
+        return True
+
+    _constraints = [
```
Comment (ThomasBinsfeld) [inline cagnotte_partner/models/cagnotte.py]: @Cedric-Pigeon `@api.constrains`

File: cagnotte_partner/models/cagnotte.py [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+                if cagnotte:
+                    values['account_cagnotte_id'] = cagnotte.id
+        return values
+
+    @api.onchange("account_cagnotte_id")
+    def onchange_account_cagnotte_id(self):
+        """ set partner on move line
+        """
+        if self.account_cagnotte_id and self.account_cagnotte_id.partner_id:
+            self.partner_id = self.account_cagnotte_id.partner_id.id
```
Comment (ThomasBinsfeld) [inline cagnotte_partner/models/cagnotte.py]: @Cedric-Pigeon `.id` is not useful

File: cagnotte_base/models/cagnotte.py [hunk 17]
```diff
@@ -3,10 +3,14 @@
 # License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
 
 from odoo import api, fields, models
+from openerp.exceptions import ValidationError
```
Comment (ThomasBinsfeld) [inline cagnotte_base/models/cagnotte.py]: @Cedric-Pigeon  s/openerp/odoo

**Comment by Cedric-Pigeon** [comment]
@JonathanNEMRY @rousseldenis @ThomasBinsfeld Code fixed following your reviews

File: cagnotte_coupon/models/cagnotte.py [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+        'A cagnotte with cagnotte type and coupon already exist'
+    )]
+
+    @api.model
+    def generate_coupon_code(self, base):
+        uid = hash(str(uuid.uuid1())) % 1000000
+        return "%s%s" % (uid, base)
+
+    @api.multi
+    def create(self, vals):
```
Comment (benwillig) [inline cagnotte_coupon/models/cagnotte.py:40]: multi decorator for a create method ?

**Comment by rousseldenis** [comment]
Isn't it time to switch those modules there https://github.com/OCA/account-payment ?

@sbidoul @Cedric-Pigeon What do you think?

**Comment by sbidoul** [comment]
@rousseldenis @Cedric-Pigeon the prerequisite to move this to OCA would be a good README that explain what these module do?

**Comment by rousseldenis** [comment]
> @rousseldenis @Cedric-Pigeon the prerequisite to move this to OCA would be a good README that explain what these module do?

Indeed, this is at this time just an idea as I'm working with those modules and thought about letting them live their life on OCA (and maybe be migrated to other versions by others :-) )

**Comment by rousseldenis** [comment]
Superseed through #147

**Pull request closed without merging**